plugins/janus_videoroom.c

```diff
@@ -987,11 +1010,15 @@ void janus_videoroom_destroy_session(janus_plugin_session *handle, int *error) {
 		if(session->participant_type == janus_videoroom_p_type_publisher) {
 			/* Get rid of publisher */
 			janus_videoroom_publisher *p = (janus_videoroom_publisher *)session->participant;
+			janus_mutex_lock(&session->mutex);
 			session->participant = NULL;
+			janus_mutex_unlock(&session->mutex);
+			janus_videoroom_leave_or_unpublish(p, TRUE);
 			if(p->room) {
+				/* Don't clear p->room.  Another thread calls janus_videoroom_leave_or_unpublish,
+					 too, and there is no mutex to protect this change. */
 				janus_refcount_decrease(&p->room->ref);
```

Comment:
But if this `janus_refcount_decrease` here frees the room, `p->room` points to garbage: isn't it better to be conservative, rather than risking `janus_videoroom_leave_or_unpublish` to crash?